lib/python/Jetson/GPIO/gpio_event.py

```diff
@@ -48,195 +66,287 @@
+#   with related information of a channel's event
+# @param[in] chip_name: the GPIO chip name/instance
+# @param[in] channel: the pin number in specified mode (board or bcm)
+# @param[out] the gpio handle with related information of a channel's event, 
+# if such handle exist, otherwise return None
+def _get_gpio_object(chip_name, channel):
+    if chip_name not in _gpio_event_list:
+        return None
+    if channel not in _gpio_event_list[chip_name]:
         return None
-    return _gpio_event_list[gpio]
+    
+    return _gpio_event_list[chip_name][channel]
 
 
 def _set_edge(gpio_name, edge):
-    edge_path = ROOT + "/" + gpio_name + "/edge"
-
-    with open(edge_path, 'w') as edge_file:
-        edge_file.write(_edge_str[edge])
+    raise RuntimeError("This function is deprecated")
 
-
-def _get_gpio_obj_key(fd):
-    for key in _gpio_event_list:
-        if _gpio_event_list[key].value_fd == fd:
-            return key
-    return None
+# @brief Look up by chip name and channel to get the gpio object 
+#   with related information of a channel's event
+# @param[in] fd: the file descriptor of a channel/line
+# @param[out] a tuple of the GPIO chip name and the pin number in 
+# specified mode, otherwise return a tuple of None
+def _get_gpio_obj_keys(fd):
+    for chip_name in _gpio_event_list:
+        for pin in _gpio_event_list[chip_name]:
+            if _gpio_event_list[chip_name][pin].value_fd == fd:
+                return (chip_name, pin)
+    return None, None
 
 
 def _get_gpio_file_object(fileno):
-    for key in _gpio_event_list:
-        if _gpio_event_list[key].value_fd.fileno() == fileno:
-            return _gpio_event_list[key].value_fd
-    return None
-
-
-def _poll_thread():
-    global _thread_running
-
-    _thread_running = True
-    while _thread_running:
+    raise RuntimeError("This function is deprecated")
+    
+# @brief A thread that catches GPIO events in a non-blocking mode.
+#   Exit upon error (may be fd non-existance, information descrepency,
+#   unknown error)
+# @param[in] thread_name: a functional name of the thread
+# @param[in] fd: the file descriptor of a channel/line
+# @param[in] channel: the pin number in specified mode (board or bcm)
+# @param[in] poll_timeout: the maximum time set to wait for edge event (second)
+def _edge_handler(thread_name, fileno, channel, poll_timeout):
+    thread_id = thread.get_ident()
+
+    # Mark the thread state as running
+    _thread_running_dict[thread_id] = True 
+
+    while _thread_running_dict[thread_id]:
         try:
-            events = _epoll_fd_thread.poll(maxevents=1)
-            fd = events[0][0]
-            _mutex.acquire()
-
-            # check if file object has been deleted or closed from main thread
-            fd = _get_gpio_file_object(fd)
-            if fd is None or fd.closed:
+            # poll for event
+            events = _epoll_fd_thread.poll(timeout=poll_timeout, maxevents=1)
```

Comment:
If we have a poll_timeout now, I think we should end the thread once the poll_timeout end